app/src/main/java/com/rkm/tasky/feature/registration/viewmodel/RegistrationViewModel.kt

```diff
@@ -0,0 +1,138 @@
+    private val passwordValidator: PasswordPatternValidator,
+    private val nameValidator: NamePatternValidator
+) : ViewModel() {
+
+    private companion object {
+        const val EMAIL_KEY = "email"
+        const val PASSWORD_KEY = "password"
+        const val NAME_KEY = "name"
+        const val SHOW_PASSWORD_KEY = "show_password"
+    }
+
+    private val mapErrors = mapOf(
+        EMAIL_KEY to R.string.registration_screen_invalid_email,
+        PASSWORD_KEY to R.string.registration_screen_invalid_password,
+        NAME_KEY to R.string.registration_screen_invalid_name
+    )
+
+    val email = TextFieldState(savedStateHandle[EMAIL_KEY] ?: "")
+    suspend fun setEmail() {
+        snapshotFlow { email.text }.collectLatest { text ->
+            savedStateHandle[EMAIL_KEY] = text
+        }
+    }
+
+    val isValidEmail by derivedStateOf {
+        emailValidator.isValidEmail(email.text.toString())
+    }
+    val password = TextFieldState(savedStateHandle[PASSWORD_KEY] ?: "")
+
+    private val isValidPassword by derivedStateOf {
+        passwordValidator.isValidPassword(password.text.toString())
+    }
+    val fullName = TextFieldState(savedStateHandle[NAME_KEY] ?: "")
+    suspend fun setName() {
+        snapshotFlow { fullName.text }.collectLatest { text ->
+            savedStateHandle[NAME_KEY] = text
+        }
+    }
+
+    val isValidName by derivedStateOf {
+        nameValidator.isValidName(fullName.text.toString())
+    }
+    private val _showPassword = MutableStateFlow(savedStateHandle[SHOW_PASSWORD_KEY] ?: false)
+    val showPassword = _showPassword.asStateFlow()
+    private val _isLoading = MutableStateFlow(false)
+    val isLoading = _isLoading.asStateFlow()
+    private val _registrationScreenEventChannel = Channel<RegistrationScreenEvent>()
+    val registrationScreenEventChannel = _registrationScreenEventChannel.receiveAsFlow()
+
+    fun onShowPasswordClicked() {
+        _showPassword.update {
+            !it
+        }
+    }
+
+    fun onRegistrationClicked() {
+        viewModelScope.launch {
+            if (!isValidName || !isValidPassword || !isValidEmail) {
+                _registrationScreenEventChannel.send(
+                    if (!isValidName) RegistrationScreenEvent.RegistrationFailedEvent(
+                        UiText.StringResource(
+                            mapErrors[NAME_KEY]!!
+                        )
+                    )
+                    else if (!isValidEmail) RegistrationScreenEvent.RegistrationFailedEvent(
+                        UiText.StringResource(
+                            mapErrors[EMAIL_KEY]!!
+                        )
+                    )
+                    else RegistrationScreenEvent.RegistrationFailedEvent(
+                        UiText.StringResource(
+                            mapErrors[PASSWORD_KEY]!!
+                        )
+                    )
+                )
+                return@launch
+            }
```

Comment:
The only thing that differs between these when branches is the key. It could therefore make sense to just retrieve that from a when expression:
```
val errorKey = when {
    !isValidName -> NAME_KEY
    !isValidPassword -> PASSWORD_KEY
    !isValidEmail -> EMAIL_KEY
    else -> null
}
if(errorKey != null) {
    _registrationScreenEventChannel.send(RegistrationFailedEvent(UiText.StringResource(mapErrors[errorKey])))
}
```